Aquality.Selenium.Core/src/Aquality.Selenium.Core/Elements/Element.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using Aquality.Selenium.Core.Applications;
+using Aquality.Selenium.Core.Elements.Interfaces;
+using Aquality.Selenium.Core.Localization;
+using Aquality.Selenium.Core.Logging;
+using Aquality.Selenium.Core.Utilities;
+using Aquality.Selenium.Core.Waitings;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Remote;
+
+namespace Aquality.Selenium.Core.Elements
+{
+    /// <summary>
+    /// Defines base class for any UI element.
+    /// </summary>
+    public abstract class Element : IElement
+    {
+        private readonly ElementState elementState;
+
+        protected Element(By locator, string name, ElementState state)
+        {
+            Locator = locator;
+            Name = name;
+            elementState = state;
+        }
+
+        public By Locator { get; }
+
+        public string Name { get; }
+
+        public virtual IElementStateProvider State => new ElementStateProvider(Locator, ConditionalWait, Finder);
+
+        protected abstract ElementActionRetrier ActionRetrier { get; }
+
+        protected abstract IApplication Application { get; }
+
+        protected abstract ConditionalWait ConditionalWait { get; }
+
+        protected abstract string ElementType { get; }
+
+        protected abstract ElementFactory Factory { get; }
+
+        protected abstract ElementFinder Finder { get; }
+
+        protected abstract LocalizationLogger LocalizationLogger { get; }
```

Comment:
maybe here also virtual + initialise with default?

Reply:
no, because it would make every element dependent of ServiceProvider (of it's particular implementation at Microsoft.Extensions.DependencyInjection). Also, we don't have a way to get ServiceProvider from here